ultradns-rest/src/main/java/denominator/ultradns/service/UltraDNSRestResourceRecordSetApi.java

```diff
@@ -0,0 +1,311 @@
+    checkNotNull(name, "name");
+    Iterator<Record> ordered = RRSetUtil.buildRecords(api
+            .getResourceRecordsOfDNameByType(zoneName, name, ResourceTypes.ALL.code())
+            .rrSets())
+            .iterator();
+    return new GroupByRecordNameAndTypeCustomIterator(ordered);
+  }
+
+  /**
+   * retrieve a resource record set by name, type.
+   *
+   * @param name      {@link ResourceRecordSet#name() name} of the rrset
+   * @param type      {@link ResourceRecordSet#type() type} of the rrset
+   * @return null unless a resource record exists with the same {@code name}, {@code type}, and
+   * {@code qualifier}
+   * @throws IllegalArgumentException if the zone is not found.
+   */
+  @Override
+  public ResourceRecordSet<?> getByNameAndType(String name, String type) {
+    checkNotNull(name, "name");
+    checkNotNull(type, "type");
+    Iterator<Record> orderedRecords = recordsByNameAndType(name, type)
+            .iterator();
+    return nextOrNull(new GroupByRecordNameAndTypeCustomIterator(
+            orderedRecords));
+  }
+
+  /**
+   * Retrieve a list of a records.
+   *
+   * @param name      {@link ResourceRecordSet#name() name} of the rrset
+   * @param type      {@link ResourceRecordSet#type() type} of the rrset
+   * @return null unless a resource record exists with the same {@code name}, {@code type}
+   * @throws IllegalArgumentException if the zone is not found.
+   */
+  private List<Record> recordsByNameAndType(String name, String type) {
+    checkNotNull(name, "name");
+    checkNotNull(type, "type");
+    int typeValue = checkNotNull(lookup(type), "typeValue for %s", type);
+    List<Record> records = null;
+    try {
+      records = RRSetUtil.buildRecords(api
+              .getResourceRecordsOfDNameByType(zoneName, name, typeValue)
+              .rrSets());
+    } catch (UltraDNSRestException e) {
+      if (e.code() == UltraDNSRestException.DATA_NOT_FOUND ||
+          e.code() == UltraDNSRestException.RESOURCE_RECORD_POOL_NOT_FOUND) {
+        records = new ArrayList<Record>();
+      } else {
+        throw e;
+      }
+    }
+    return records;
+  }
+
+  /**
+   * This method will update Records for a combination
+   * of owner, resource type & group.
+   * @param rrset If {@link
+   *              ResourceRecordSet#ttl() ttl} is not present, zone default is used.
+   */
+  @Override
+  public void put(ResourceRecordSet<?> rrset) {
+    checkNotNull(rrset, "rrset was null");
+    checkArgument(!rrset.records().isEmpty(), "rrset was empty %s", rrset);
+    int ttlToApply = rrset.ttl() != null ? rrset.ttl() : DEFAULT_TTL;
+    int typeCode = lookup(rrset.type());
+
+    RRSetList rrSetList = getRrSetList(rrset.name(), typeCode);
+
+    if (rrSetList == null || rrSetList.getRrSets() == null) {
+      // Create the resource record set.
+      // If the zone does not exist, should it be created? I think it should not be created. Check the SOAP behaviour.
```

Comment:
Please be certain what your code will do & modify the comment accordingly & clearly. This is an open source project & people can read your comments.